### ISSUE TITLE
Spire Agent Init Container Edge Case

### ISSUE THREAD

**Issue by clhain** [nginxinc/nginx-service-mesh#81]
Hey there,
I'm running from the latest 0.5.0 helm release, and am running into issues with spire-agent getting stuck in init phase following spire-server exiting with 'Completed' status:

```
$ kubectl get pods
NAME                                  READY   STATUS      RESTARTS       AGE
nats-server-865747449-xffnh           2/2     Running     0              145m
nginx-mesh-api-696555f7d6-hmdwx       0/1     Completed   0              25h
nginx-mesh-api-696555f7d6-k75wc       1/1     Running     9 (113m ago)   145m
nginx-mesh-metrics-8448884887-gcr45   1/1     Running     6 (115m ago)   145m
nginx-mesh-metrics-8448884887-n7hjb   0/1     Completed   0              25h
spire-agent-8n5jr                     1/1     Running     1 (112m ago)   20h
spire-agent-dls89                     1/1     Running     1 (112m ago)   17h
spire-agent-rjcd9                     0/1     Init:0/1    0              142m
spire-server-5646f7f9d9-p2tkl         2/2     Running     0              145m
spire-server-5646f7f9d9-vq57t         0/2     Completed   0              25h
```

```
$ kubectl logs -c init  spire-agent-rjcd9
E1021 20:31:55.081942   41641 reflector.go:140] vendor/k8s.io/client-go/tools/watch/informerwatcher.go:146: Failed to watch *unstructured.Unstructured: unknown
error: timed out waiting for the condition on pods/spire-server-5646f7f9d9-vq57t
...
```

This is the command the init container is executing, run manually:
```
$ kubectl -n nginx-mesh wait --for=condition=Ready pod -l app.kubernetes.io/name=spire-server > /dev/null
error: timed out waiting for the condition on pods/spire-server-5646f7f9d9-vq57t
```

Here's the message from the spire-server container in Completed state:
```
$ kubectl describe pod spire-server-5646f7f9d9-vq57t | grep -A2 Status:
Status:       Failed
Reason:       Terminated
Message:      Pod was terminated in response to imminent node shutdown.
```

Deleting the failed container allows the pod to recover.
```
$ kubectl get pod spire-agent-rjcd9
NAME                READY   STATUS     RESTARTS   AGE
spire-agent-rjcd9   0/1     Init:0/1   0          150m

$ kubectl delete pod spire-server-5646f7f9d9-vq57t
pod "spire-server-5646f7f9d9-vq57t" deleted

$ kubectl get pod spire-agent-rjcd9
NAME                READY   STATUS    RESTARTS   AGE
spire-agent-rjcd9   1/1     Running   0          155m
```

Suggest updating the init container to either watch for an active endpoint on the service, or some other active check against the service rather than the individual pods.

**Comment by sjberman** [comment]
Thanks @clhain , we have a change to the init container coming in our next release which should hopefully remedy this issue.

**Comment by sjberman** [comment]
This issue should now be resolved with the most recent 1.6 release.

**Comment by ThomasJeffers** [comment]
Hi, I've created a k8s cluster with one control plane and one worker node with kubeadm and tried to install Spire

and encountered this error in which spire-agent pod won't get created

```
NAMESPACE     NAME                               READY   STATUS     RESTARTS         AGE
kube-system   cilium-66ls7                       1/1     Running    1 (8h ago)       8h
kube-system   cilium-operator-6d7886cdf4-nlfjg   1/1     Running    1 (8h ago)       8h
kube-system   cilium-vrnqr                       1/1     Running    0                8h
kube-system   coredns-787d4945fb-4gfsx           1/1     Running    1 (8h ago)       18h
kube-system   coredns-787d4945fb-5ht99           1/1     Running    1 (8h ago)       18h
kube-system   etcd-root                          1/1     Running    5 (8h ago)       18h
kube-system   kube-apiserver-root                1/1     Running    5 (7h59m ago)    18h
kube-system   kube-controller-manager-root       1/1     Running    1 (8h ago)       18h
kube-system   kube-proxy-7lvxr                   1/1     Running    0                8h
kube-system   kube-proxy-bxkq2                   1/1     Running    1 (8h ago)       18h
kube-system   kube-scheduler-root                1/1     Running    1 (8h ago)       18h
spire         spire-agent-chjmd                  0/1     Init:0/2   49 (5m35s ago)   4h10m
spire         spire-server-0                     2/2     Running    0                4h10m
```

I'm desperate to know what's wrong. bear in mind that I am novice to Kubernetes world.
